pkg/scheduler/queue/tree_queueing_algorithms.go

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+// DequeueAlgorithm represents the set of operations specific to different approaches to dequeuing. It is applied
+// at the layer-level -- every Node at the same depth in a TreeQueue shares the same DequeueAlgorithm, including state
+// that may be stored in a struct that implements DequeueAlgorithm.
+type DequeueAlgorithm interface {
+	addChildNode(parent, child *Node)
+	dequeueSelectNode(dequeueFrom *Node) (*Node, bool)
+	dequeueUpdateState(dequeueParent *Node, dequeuedFrom *Node)
```

Comment:
I think we still have an impedance mismatch with arg naming here as the `dequeueFrom *Node` is the parent from which we select the child, but afterwards when we update state, the `dequeuedFrom *Node` is the child.

I think it's a bit less confusing if we drop the "parent" term because we are being handed a node and we never look at _that_ node's parent, only it's child. Having "parent node" "node" and "child node" terms in play implies three levels but we're only ever operating at two levels - current and child. 
possibly just

 ```golang
dequeueSelectNode(node *Node) (*Node, bool)
dequeueUpdateState(node *Node, dequeuedFrom *Node)
```